src/contracts/deposit/deposit.service.ts

```diff
@@ -11,50 +11,53 @@ import {
 } from './deposit.constants';
 import {
   DepositEvent,
-  VerifiedDepositEvent,
   VerifiedDepositEventsCache,
-  VerifiedDepositEventsCacheHeaders,
   VerifiedDepositEventGroup,
 } from './interfaces';
 import { RepositoryService } from 'contracts/repository';
-import { CacheService } from 'cache';
 import { BlockTag } from 'provider';
 import { BlsService } from 'bls';
 import { APP_VERSION } from 'app.constants';
+import { DepositIntegrityCheckerService } from './integrity-checker';
+import { parseLittleEndian64 } from './deposit.utils';
+import { DepositTree } from './deposit-tree';
+import { LevelDBService } from './leveldb';
 
 @Injectable()
 export class DepositService {
   constructor(
     @Inject(WINSTON_MODULE_NEST_PROVIDER) private logger: LoggerService,
     private providerService: ProviderService,
     private repositoryService: RepositoryService,
-    private cacheService: CacheService<
-      VerifiedDepositEventsCacheHeaders,
-      VerifiedDepositEvent
-    >,
+
     private blsService: BlsService,
+    private depositIntegrityCheckerService: DepositIntegrityCheckerService,
+    private levelDBCacheService: LevelDBService,
   ) {}
 
   public async handleNewBlock(blockNumber: number): Promise<void> {
     if (blockNumber % DEPOSIT_EVENTS_CACHE_UPDATE_BLOCK_RATE !== 0) return;
 
     // The event cache is stored with an N block lag to avoid caching data from uncle blocks
     // so we don't worry about blockHash here
-    await this.updateEventsCache();
+    const toBlockNumber = await this.updateEventsCache();
+    await this.depositIntegrityCheckerService.checkFinalizedRoot(toBlockNumber);
   }
 
   public async initialize(blockNumber: number) {
```

Comment:
here we fetch cache twice: for validation of cache age and for update 
is it make sence to send cache header as parameter to updateEventsCache() ?

Reply:
I'm not sure what you mean
Could you explain?

Reply:
I just wanted to highlight that we fetch events cache twice in `initialize` method.

Reply:
it's not a big problem because we're pulling events from the local database. It can be optimized, but it won't give much change at startup, because the query to the node takes many times more time
But if we do decide to do that, we'll have to complicate the code, which we don't want to do